File: node/src/tree/tree.js
// enum class for traversal and processing order
var TraverseMethod = {
   PRE_ORDER: 1,
   IN_ORDER: 2,
   POST_ORDER: 3
}

// enum class for comparisons, gt, lt and equal
var ComparisonSign = {
    LESS: 1,
    GREATER: 2,
    EQUAL: 3
}

// class for nodes within Tree
class Node {
   constructor(value) {
      this.value = value
      this.left = null
      this.right = null
   }
}

// BinaryTree Implementation
class BinaryTree {

   constructor() {
      this.root = null
   }

   traverse(method, action_func) {

      function _visit(node) {
         if (method == TraverseMethod.PRE_ORDER) {
            action_func(node.value)
         }

         if (node.left) {
            _visit(node.left)
         }

         if (method == TraverseMethod.IN_ORDER) {
            action_func(node.value)
         }

         if (node.right) {
            _visit(node.right)
         }

         if (method == TraverseMethod.POST_ORDER) {
            action_func(node.value)
         }
      }

      _visit(this.root)
   }

   returnAsArr(method) {
      let result = []

      this.traverse(method, function(value) {
         result.push(value)
      })

      return result
   }
}

class BinarySearchTree extends BinaryTree {

   constructor(comparison_func=null) {
      super()

      if (comparison_func == null) {
         this.comparison_func = this.comparison_func_default
      } else {
         this.comparison_func = comparison_func
      }
   }

   //  default comparison function
   comparison_func_default(val1, val2, CS) {
      if (CS == ComparisonSign.EQUAL) {
         return val1 == val2
      }
      else
      if (CS == ComparisonSign.LESS) {
         return val1 < val2
      }
      else
      if (CS == ComparisonSign.GREATER) {
         return val1 > val2
      }
      else {
         return false  // Never get here
      }
   }

   // adds new value to the tree
   add(new_value) {

      // recursive method for evaluating a node and calling itself depending on the value
      // @TO-RESEARCH:  see how an arrow-function is used to access the "this", learn more!
      var _find_and_insert = (node) => {

         if (this.comparison_func(node.value, new_value, ComparisonSign.GREATER))
               if (node.left == null) {
                  node.left = new Node(new_value)
               } else {
                  _find_and_insert(node.left)
               }

         if (this.comparison_func(node.value, new_value, ComparisonSign.LESS))
               if (node.right == null) {
                  node.right = new Node(new_value)
               } else {
                  _find_and_insert(node.right)
               }
      }

      if (this.root == null) {
         this.root = new Node(new_value)
      } else {
         _find_and_insert(this.root)
      }
   }

   // accepts a value, and returns a boolean indicating whether or not the value is in the tree at least once.
   contains(target_value) {

      // recursive function for isiting each node
      // @TO-RESEARCH:  see how an arrow-function is used to access the "this", learn more!
      var _visit = (node) => {
         if (this.comparison_func(node.value, target_value, ComparisonSign.EQUAL))
               return true
         if (this.comparison_func(node.value, target_value, ComparisonSign.GREATER))
               if (node.left == null) {
                  return false
               } else {
                  _visit(node.left)
               }
         if (this.comparison_func(node.value, target_value, ComparisonSign.LESS))
               if (node.right == null) {
                  return false
               } else {
                  _visit(node.right)
               }
      }

      if (this.root != null) {
         return _visit(this.root)
      } else {
         return false
      }

   }

}

module.exports.Node = Node
module.exports.BinaryTree = BinaryTree
module.exports.BinarySearchTree = BinarySearchTree
module.exports.TraverseMethod = TraverseMethod
module.exports.ComparisonSign = ComparisonSign

File: golang/src/tree/tree.go
package tree

import (
	"fmt"
	"os"
)

const (
	// TraversalOrderPre left->middle->right order
	TraversalOrderPre = iota
	// TraversalOrderIn middle->left->right order
	TraversalOrderIn = iota
	// TraversalOrderPost left->right->middle order
	TraversalOrderPost = iota
)

const (
	// ComparisonSignLess is the literal Less-Than Sign
	ComparisonSignLess = iota
	// ComparisonSignGreater is the literal Greater-Than Sign
	ComparisonSignGreater = iota
	// ComparisonSignEqual is the literal Equal Sign
	ComparisonSignEqual = iota
)

// Node for Tree
type Node struct {
	Val   interface{}
	left  *Node
	right *Node
}

// NewNode constructor for Node
func NewNode(val interface{}) Node {
	return Node{val, nil, nil}
}

// BinarySearchTree implementation
type BinarySearchTree struct {
	root           *Node
	comparisonFunc func(val1 interface{}, val2 interface{}, CS int) bool
}

// NewBinarySearchTree of Binary Tree
func NewBinarySearchTree(comparisonFunc func(val1 interface{}, val2 interface{}, CS int) bool) BinarySearchTree {
	t := BinarySearchTree{nil, nil}
	if comparisonFunc == nil {
		t.comparisonFunc = t.comparisonFuncDefault
	} else {
		t.comparisonFunc = comparisonFunc
	}
	return t
}

// Traverse of BinarySearchTree
func (tree *BinarySearchTree) Traverse(method int, actionFunc func(val interface{})) {

	// formally defining the var to the func, will allow us to call-ourselves
	var _visit func(node *Node)

	_visit = func(node *Node) {
		if method == TraversalOrderPre {
			actionFunc(node.Val)
		}

		if node.left != nil {
			_visit(node.left)
		}

		if method == TraversalOrderIn {
			actionFunc(node.Val)
		}

		if node.right != nil {
			_visit(node.right)
		}

		if method == TraversalOrderPost {
			actionFunc(node.Val)
		}
	}

	if tree.root != nil {
		_visit(tree.root)
	}

}

// ReturnAsArr will return the tree as as array
func (tree *BinarySearchTree) ReturnAsArr(method int) []interface{} {

	var retval []interface{} = make([]interface{}, 0)

	actionFunc := func(value interface{}) {
		retval = append(retval, value)
	}

	tree.Traverse(method, actionFunc)

	return retval
}

// comparison_func_default is the default of comparing values
func (tree *BinarySearchTree) comparisonFuncDefault(val1 interface{}, val2 interface{}, CS int) bool {
	if CS == ComparisonSignEqual {
		return val1.(string) == val2.(string)
	} else if CS == ComparisonSignLess {
		return val1.(string) < val2.(string)
	} else if CS == ComparisonSignGreater {
		return val1.(string) > val2.(string)
	} else {
		return false
	}
}

// findAndInsert recursive method for evaluating a node and calling itself depending on the value
func (tree *BinarySearchTree) findAndInsert(node *Node, newValue interface{}) {

	if tree.comparisonFunc == nil {
		fmt.Printf("ERROR! tree.comparisonFunc==nil")
		os.Exit(99)
	}

	if tree.comparisonFunc(node.Val, newValue, ComparisonSignGreater) {
		if node.left == nil {
			node.left = &Node{newValue, nil, nil}
		} else {
			tree.findAndInsert(node.left, newValue)
		}
	}

	if tree.comparisonFunc(node.Val, newValue, ComparisonSignLess) {
		if node.right == nil {
			node.right = &Node{newValue, nil, nil}
		} else {
			tree.findAndInsert(node.right, newValue)
		}
	}
}

// Add a value to the Tree
func (tree *BinarySearchTree) Add(newValue interface{}) {

	if tree.root == nil {
		node := Node{newValue, (*Node)(nil), (*Node)(nil)}
		tree.root = &node
	} else {
		tree.findAndInsert(tree.root, newValue)
	}

}

// Contains accepts a value, and returns a boolean indicating whether or not the value is in the tree at least once.
func (tree *BinarySearchTree) Contains(targetVal interface{}) bool {

	// *** Return Early
	if tree.root == nil {
		return false
	}

	if tree.comparisonFunc == nil {
		fmt.Printf("ERROR! tree.comparisonFunc==nil")
		os.Exit(99)
	}

	var _visit func(node *Node) bool

	_visit = func(node *Node) bool {

		if tree.comparisonFunc(node.Val, targetVal, ComparisonSignEqual) {
			return true
		}
		if tree.comparisonFunc(node.Val, targetVal, ComparisonSignGreater) {
			if node.left == nil {
				return false
			}
			_visit(node.left)
		}
		if tree.comparisonFunc(node.Val, targetVal, ComparisonSignLess) {
			if node.right == nil {
				return false
			}
			_visit(node.right)
		}
		return false
	}

	return _visit(tree.root)
}
